sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaExactlyOnceSink.java

```diff
@@ -524,6 +526,20 @@ void commitTxn(long lastRecordId, Counter numTransactions) throws IOException {
           ProducerSpEL.commitTransaction(producer);
 
           numTransactions.inc();
+          if (!reportedLineage) {
+            Lineage.getSinks()
+                .add(
+                    "kafka",
+                    ImmutableList.of(
```

Comment:
Do we need to sort and/or deduplicate the content of the collected lineage in this list to reduce the number different lineage sources?

Reply:
The "ImmutableList" is a list of segments, the `.add(String system, List segments)` call reports one lineage string. I did some refactoring in #32090 in order to handle special characters in segments, see https://github.com/apache/beam/pull/32090#issuecomment-2274138958

-----

If the question is asking about report lineage multiple times,  here the reportedLineage flag is a simple way to deduplicate reporting same metrics too often. Because Lineage is a Set, report multiple times does not harm, just adds some overhead.

Here each ShardWriter instance always run on same topic (spec is a constructor argument, and the topic it writes is `spec.getTopic()`) so a boolean flag suffices.

Reply:
> Because Lineage is a Set, report multiple times does not harm

Ok, great, I think this is what I was asking about.